lightningd/pay.c

```diff
@@ -209,12 +263,26 @@ void payment_failed(struct lightningd *ld, const struct htlc_out *hout,
 	wallet_payment_set_status(ld->wallet, &hout->payment_hash,
 				  PAYMENT_FAILED, NULL);
 
-	/* FIXME: save ids we can turn reply->origin_index into sender. */
+	/* Report to gossipd if there is something we can report. */
+	if (fail) {
+		log_debug(ld->log,
+			  "Reporting route failure to gossipd: 0x%04x (%s) "
+			  "node %s channel %s update %s",
+			  fail->failcode, onion_type_name(fail->failcode),
+			  type_to_string(tmpctx, struct pubkey,
+					 &fail->erring_node),
+			  type_to_string(tmpctx, struct short_channel_id,
+				  	 &fail->erring_channel),
+			  tal_hex(tmpctx, fail->channel_update));
+		report_routing_failure(ld->gossip, fail);
+	}
 
-	/* FIXME: check for routing failure / perm fail. */
-	/* check_for_routing_failure(i, sender, failure_code); */
+	/* FIXME(ZmnSCPxj): if retrying is plausible, and we are
+	 * using pay command rather than sendpay, retry routing
+	 * and payment again. */
+	(void) retry_plausible;
 
-	json_pay_failed(hout->cmd, NULL, failcode, "reply from remote");
+	json_pay_failed(hout->cmd, failcode, failmsg);
```

Comment:
Future work: we should provide more detailed failmsg: unlike logging, that's returned to the caller who is definitely interested!

Reply:
Issue to track this: #866